refactor(test): improve L1 files L1CrossDomainMessenger and OptimismPortal2 test coverage and quality with AI assistance [Prompt v0.7.1]

aliersh opened the pull request:
## Overview

This PR introduces AI-guided test improvements for both `L1CrossDomainMessenger` and `OptimismPortal2`, using the newly revised [Prompt v0.7.1](https://www.notion.so/oplabs/Prompt-v0-7-1-239f153ee16280c294a2e483de38e91d?source=copy_link). It replaces [Prompt v0.7.0](https://www.notion.so/oplabs/Prompt-v0-7-0-237f153ee162809f9d2df76ee831bc34?source=copy_link), which generated inconsistent and non-actionable outputs in previous test generation efforts.

Prompt v0.7.1 introduces critical upgrades to AI-generated test reliability, including:

- Strict constraints enforcement (e.g., no redundant `vm.assume`)
- Accurate alignment with contract behavior and function requirements
- Mandatory heavy fuzz validation
- Output standardization and explicit coverage expectations

Check the [prompt changelog](https://www.notion.so/oplabs/Prompt-changelog-225f153ee1628010990ef3dc24cb0b83?source=copy_link) for more information.

This version ensures deterministic, spec-aligned test generation that emphasizes:

1. Converting standard tests into meaningful fuzz tests
2. Covering all observable code paths
3. Ensuring every public/external function has at least one targeted test

All tests in this PR were generated automatically via Prompt v0.7.1, without manual editing or validation intervention.

## L1CrossDomainMessenger Test Improvements

### New and Enhanced Tests

- `test_portal_succeeds`: Validates that the legacy `PORTAL()` getter correctly returns the assigned portal address.
- `testFuzz_initialize_anySystemConfig_succeeds`: Fuzz test that verifies `initialize()` correctly handles arbitrary system config addresses.
- `testFuzz_initialize_anyPortal_succeeds`: Fuzz test that checks `initialize()` accepts arbitrary portal addresses and sets them properly.
- `testFuzz_upgrade_anySystemConfig_succeeds`: Fuzz test validating that `upgrade()` can be called with any valid system config address and that the internal state is updated accordingly.
- `testFuzz_sendMessage_varyingInputs_succeeds`: Fuzzes `sendMessage()` with randomized inputs for gas limit, message payload, and sender.
- `test_sendMessage_zeroGasLimit_succeeds`: Confirms that `sendMessage()` behaves correctly when passed a zero gas limit (i.e., does not revert unexpectedly).
- `test_sendMessage_highGasLimit_reverts`: Verifies that `sendMessage()` reverts gracefully when passed an unreasonably high gas limit that would exhaust available gas.

### Modified Tests

- `testFuzz_relayMessage_varyingInputs_succeeds`: Converted from a regular test to a fuzz test covering target, gas, and payload.
- `test_xDomainMessageSender_duringSend_reverts`: Refactored and simplified; no longer marked as fuzz test.

### New Targeted Test Contracts

- `L1CrossDomainMessenger_Portal_Test`: Isolated test contract for legacy getter `PORTAL()`.

### Test Count

- Increased from 35 → 42
  All tests pass cleanly.

## OptimismPortal2 Test Improvements

### New Tests

- Getter tests:
  - `test_paused_succeeds`
  - `test_proofMaturityDelaySeconds_succeeds`
  - `test_disputeGameFactory_succeeds`
  - `test_superchainConfig_succeeds`
  - `test_guardian_succeeds`
  - `test_disputeGameFinalityDelaySeconds_succeeds`
  - `test_respectedGameType_succeeds`
  - `test_respectedGameTypeUpdatedAt_succeeds`

- Dispute-related tests:
  - `test_disputeGameBlacklist_nonBlacklisted_succeeds`
  - `testFuzz_disputeGameBlacklist_succeeds`

- Proof submitter tracking:
  - `test_numProofSubmitters_unprovenWithdrawal_succeeds`
  - `test_numProofSubmitters_provenWithdrawal_succeeds`
  - `testFuzz_numProofSubmitters_multipleProofs_succeeds`

### Converted and Split Tests

- `test_minimumGasLimit_succeeds` was split into:
  - `testFuzz_minimumGasLimit_increasesLinearly_succeeds`: A fuzz test with bounded linear constraints
  - `test_minimumGasLimit_zeroCalldata_succeeds`: Covers edge case with zero calldata

### Structural Enhancements

- Improved `MinimumGasLimit_Test` contract structure
- Refined naming and separation between fuzz and non-fuzz logic

### Test Count

- Increased from 78 → 91  
  All tests pass with full CI and heavy fuzz test success.

---

## Impact

- 20 new tests added across both files
- 3 existing tests converted into fuzz tests
- Getter, edge case, and behavioral coverage expanded significantly
- Generated using Prompt v0.7.1 with no manual adjustments
- All tests pass and meet project and prompt-defined validation standards

---

## Manual Changes

While the majority of this PR was generated via Prompt v0.7.1, a few targeted manual changes were made based on post-generation review. These will be incorporated into future prompt improvements to eliminate the need for manual intervention.

### Converted Tests

Several tests that previously focused on zero address edge cases were converted into fuzz tests to ensure broader behavioral coverage:

- `test_initialize_zeroSystemConfig_succeeds`  
  → replaced with `testFuzz_initialize_anySystemConfig_succeeds`

- `test_initialize_zeroPortal_succeeds`  
  → replaced with `testFuzz_initialize_anyPortal_succeeds`

- `test_upgrade_zeroSystemConfig_succeeds`  
  → replaced with `testFuzz_upgrade_anySystemConfig_succeeds`

These changes improve test quality by demonstrating that these functions accept any address, not just a specific edge case. They also align with the prompt's goal of exercising realistic inputs without artificially constraining the test logic.

### Removed Test

- `test_pause_zeroSystemConfig_reverts` was removed

This test exercised an invalid contract setup (zero `systemConfig` address) that does not reflect supported or expected behavior. Since the scenario falls outside meaningful improvement, the test was removed.

### Prompt Update Note

The need for these changes reflects areas that will be addressed in the next prompt version. Specifically:
- Fuzz test generation will default to broad input coverage instead of fixed-value edge cases
- Tests for invalid or unsupported setup states will be filtered out to avoid speculative behavior

aliersh committed:
refactor(test): enhance L1CrossDomainMessenger test coverage with fuzz tests and edge cases

  - Add test for PORTAL() legacy getter function
  - Add edge case tests for zero address handling in initialize() and upgrade()
  - Add boundary tests for sendMessage with zero and high gas limits
  - Convert regular tests to fuzz tests for better randomized coverage
  - Add paused() state test when systemConfig is zero address
  - Increase test count from 35 to 43 with 100% function coverage

aliersh committed:
refactor(test): improve OptimismPortal2 test coverage and quality

- add 13 tests for uncovered getter functions
- convert 1 test to fuzz test with proper bounds

aliersh committed:
fix(test): specify empty revert reason for zero address call in L1CrossDomainMessenger test

smartcontracts commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -128,6 +128,40 @@ contract L1CrossDomainMessenger_Initialize_Test is L1CrossDomainMessenger_TestIn
     }
+
+    /// @notice Tests that initialize handles zero addresses correctly.
+    function test_initialize_zeroSystemConfig_succeeds() external {
```
This is an interesting case. Technically it's a behavior we're OK with, in theory, since we don't really do a lot of validation on inputs, but it's also strange to test that it's allowed. I guess for now we should permit the test to show that this is technically allowed.

smartcontracts replied:
A more accurate test here would be turning this into a fuzz test, where instead of specifically having zero, it just uses any address and demonstrates that you can input any address and it won't reject that input.

aliersh replied:
Fixed in [a7133ec](https://github.com/ethereum-optimism/optimism/pull/16812/commits/a7133ec6a70a60483da3dcf82199efaa02d9ec5c).
Converted to a fuzz test.

smartcontracts commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -128,6 +128,40 @@ contract L1CrossDomainMessenger_Initialize_Test is L1CrossDomainMessenger_TestIn
+    }
+
+    /// @notice Tests that initialize handles zero portal address correctly.
+    function test_initialize_zeroPortal_succeeds() external {
```
Same as above, testing for the zero value is weird, it's much better if we just cover the input with a fuzz test

aliersh replied:
Fixed in [a7133ec](https://github.com/ethereum-optimism/optimism/pull/16812/commits/a7133ec6a70a60483da3dcf82199efaa02d9ec5c).
Converted to a fuzz test.

smartcontracts commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -203,6 +237,22 @@ contract L1CrossDomainMessenger_Upgrade_Test is L1CrossDomainMessenger_TestInit
         l1CrossDomainMessenger.upgrade(newSystemConfig);
     }
+
+    /// @notice Tests that upgrade handles zero address for system config.
```
Same as above

aliersh replied:
Fixed in [a7133ec](https://github.com/ethereum-optimism/optimism/pull/16812/commits/a7133ec6a70a60483da3dcf82199efaa02d9ec5c).
Converted to a fuzz test.

smartcontracts commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -228,6 +278,20 @@ contract L1CrossDomainMessenger_Paused_Test is L1CrossDomainMessenger_TestInit {
     }
+
+    /// @notice Tests paused state when systemConfig is zero address.
+    function test_pause_zeroSystemConfig_reverts() external {
```
Another weird test, this isn't a behavior we actually want, it just shows that if you provide a system config that isn't actually a system config, then any functions that we depend on will fail. If we accept this test then there are a million other tests we should be writing. We probably need to tweak the prompt a bit so it doesn't go down this route.

aliersh replied:
Fixed in [a7133ec](https://github.com/ethereum-optimism/optimism/pull/16812/commits/a7133ec6a70a60483da3dcf82199efaa02d9ec5c).
Test removed.

smartcontracts reviewed (commented):
Left some feedback on a few tests, the remainder look good to me.

aliersh committed:
fix(test): refactor L1CrossDomainMessenger tests to use fuzz testing

- convert test_initialize_zeroSystemConfig_succeeds to testFuzz_initialize_anySystemConfig_succeeds
- convert test_initialize_zeroPortal_succeeds to testFuzz_initialize_anyPortal_succeeds
- convert test_upgrade_zeroSystemConfig_succeeds to testFuzz_upgrade_anySystemConfig_succeeds
- remove test_pause_zeroSystemConfig_reverts as it tests invalid setup

smartcontracts commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -203,6 +239,23 @@ contract L1CrossDomainMessenger_Upgrade_Test is L1CrossDomainMessenger_TestInit
+
+    /// @notice Fuzz test for upgrade with any system config address.
+    /// @param _systemConfig The system config address to test.
+    function testFuzz_upgrade_anySystemConfig_succeeds(address _systemConfig) external {
```
Let's remove this test, not because it's bad but because it'll create a merge conflict with another PR

aliersh replied:
Fixed in [20ef1d4](https://github.com/ethereum-optimism/optimism/pull/16812/commits/20ef1d434734cab1b1bdc8a02c9eb0ca233ef8fd)

smartcontracts reviewed (commented):
Mostly good, one comment!

aliersh committed:
fix(test): remove testFuzz_upgrade_anySystemConfig_succeeds to avoid merge conflict

JosepBove commented:
After Kelvin's suggestions it looks perfect
